src/core/expression/qgsexpressionfunction.cpp

```diff
@@ -6730,6 +6750,126 @@ static QVariant executeGeomOverlay( const QVariantList &values, const QgsExpress
+        {
+          case QgsWkbTypes::GeometryType::PolygonGeometry:
+          {
+            // Check min overlap for intersection (if set)
+            if ( intersection.isMultipart() )
+            {
+              bool testResult { false };
+              for ( auto it = intersection.const_parts_begin(); ! testResult && it != intersection.const_parts_end(); ++it )
+              {
+                const QgsPolygon *geom = qgsgeometry_cast< const QgsPolygon * >( *it );
+                // qDebug() << "Area" << feat2.id() << geom->area();
+                if ( minOverlap != -1 )
+                {
+                  if ( geom->area() >= minOverlap )
+                  {
+                    testResult = true;
+                  }
+                  else
+                  {
+                    continue;
+                  }
+                }
+
+                // Check min inscribed circle radius for intersection (if set)
+                if ( minInscribedCircleRadius != -1 )
+                {
+                  const QgsRectangle bbox = geom->boundingBox();
+                  const double width = bbox.width();
+                  const double height = bbox.height();
+                  const double size = width > height ? width : height;
+                  const double tolerance = size / 1000.0;
+                  //qDebug() << "Inscribed circle radius" << feat2.id() << QgsGeos( geom ).maximumInscribedCircle( tolerance )->length();
+                  testResult = QgsGeos( geom ).maximumInscribedCircle( tolerance )->length() >= minInscribedCircleRadius;
+                }
+              }
+
+              if ( ! testResult )
+              {
+                continue;
+              }
+
+            }
+            else
+            {
+              if ( minOverlap != -1 && intersection.area() < minOverlap )
+              {
+                continue;
+              }
+
+              // Check min inscribed circle radius for intersection (if set)
+              if ( minInscribedCircleRadius != -1 )
+              {
+                const QgsAbstractGeometry *geom { intersection.constGet() };
+                const QgsRectangle bbox = geom->boundingBox();
+                const double width = bbox.width();
+                const double height = bbox.height();
+                const double size = width > height ? width : height;
+                const double tolerance = size / 1000.0;
+                // qDebug() << "Inscribed circle radius" << feat2.id() << QgsGeos( geom ).maximumInscribedCircle( tolerance )->length();
+                if ( QgsGeos( geom ).maximumInscribedCircle( tolerance )->length() < minInscribedCircleRadius )
+                {
+                  continue;
+                }
+              }
+            }
+            break;
+          }
+          case QgsWkbTypes::GeometryType::LineGeometry:
+          {
+            // Check min overlap for intersection (if set)
+            if ( intersection.isMultipart() )
```

Comment:
again, we can safely remove the two branches here and always use the multipart approach